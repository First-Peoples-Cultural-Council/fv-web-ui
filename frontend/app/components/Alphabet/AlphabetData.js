import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import selectn from 'selectn'

// FPCC
import ProviderHelpers from 'common/ProviderHelpers'
import { SECTIONS } from 'common/Constants'

// DataSources
import useCharacters from 'dataSources/useCharacters'
import useIntl from 'dataSources/useIntl'
import useProperties from 'dataSources/useProperties'
import useRoute from 'dataSources/useRoute'
import useWindowPath from 'dataSources/useWindowPath'

/**
 * @summary AlphabetData
 * @version 1.0.1
 * @component
 *
 * @param {object} props
 * @param {function} props.children
 *
 */
function AlphabetData({ children, dialect }) {
  const { computeCharacters, fetchCharacters } = useCharacters()
  const { intl } = useIntl()
  const { properties } = useProperties()
  const { routeParams } = useRoute()
  const { pushWindowPath, splitWindowPath } = useWindowPath()

  const fetchData = () => {
    fetchCharacters(routeParams.dialect_path + '/Alphabet', '&sortOrder=asc&sortBy=fvcharacter:alphabet_order')
  }

  useEffect(() => {
    fetchData()
  }, [])

  const _computeCharacters = ProviderHelpers.getEntry(computeCharacters, routeParams.dialect_path + '/Alphabet')

  const rawCharacters = selectn('response.entries', _computeCharacters)
  const dialectName = selectn('response.title', dialect.compute)

  const [currentChar, setCurrentChar] = useState({})

  const characters = rawCharacters
    ? rawCharacters.map((char) => {
        return {
          uid: char.uid,
          title: char.title,
          audio: selectn('contextParameters.character.related_audio[0].path', char),
          path: char.path,
        }
      })
    : null

  const onCharacterClick = (character) => {
    const charElement = document.getElementById('charAudio' + character.uid)
    if (charElement) {
      document.getElementById('charAudio' + character.uid).play()
    }
    setCurrentChar(character)
  }

  const onCharacterLinkClick = () => {
    const newPathArray = splitWindowPath.slice()
    newPathArray.push(currentChar.title)
    pushWindowPath('/' + newPathArray.join('/'))
  }

  return children({
    characters,
    currentChar,
    dialectName,
    intl,
    isSections: routeParams.area === SECTIONS,
    onCharacterClick,
    onCharacterLinkClick,
    properties,
  })
}
// PROPTYPES
const { func } = PropTypes
AlphabetData.propTypes = {
  children: func,
}

export default AlphabetData
