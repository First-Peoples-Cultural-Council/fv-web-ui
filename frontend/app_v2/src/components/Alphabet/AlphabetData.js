import { useEffect, useState } from 'react'
import api from 'services/api'
import getAlphabetAdaptor from 'services/api/adaptors/getAlphabet'
import useGetSections from 'common/useGetSections'
import { useParams } from 'react-router-dom'
/**
 * @summary AlphabetData
 * @component NB: This component is used by multiple Presentation layers
 *
 * @param {object} props
 *
 */
export const findSelectedCharacterData = ({ character, data, language }) => {
  const characters = data?.characters
  const found = characters.find(({ title }) => title === character)
  if (found?.relatedEntries) {
    found.relatedEntries.forEach((entry) => {
      entry.url = `/${language}/word/${entry.uid}`
    })
  }
  return found
}

const AlphabetData = () => {
  const { title } = useGetSections()
  const { character, language } = useParams()
  const { isLoading, error, data } = api.getAlphabet(title, getAlphabetAdaptor)
  const [selectedData, setSelectedData] = useState({})

  useEffect(() => {
    if (character && data) {
      const _selectedData = findSelectedCharacterData({ character, data, language })
      if (_selectedData !== undefined && _selectedData?.title !== selectedData?.title) {
        setSelectedData(_selectedData)
      }
    }
  }, [character, data])

  // Video Modal
  const [videoIsOpen, setVideoIsOpen] = useState(false)

  const onCharacterClick = (clickedCharacter) => {
    const _selectedData = findSelectedCharacterData({ character: clickedCharacter, data, language })
    if (_selectedData !== undefined && _selectedData?.title !== selectedData?.title) {
      setSelectedData(_selectedData)
    }
  }

  const onVideoClick = () => {
    setVideoIsOpen(!videoIsOpen)
  }

  return {
    characters: data?.characters,
    links: data?.links,
    error,
    isLoading,
    language,
    onCharacterClick,
    onVideoClick,
    selectedData,
    videoIsOpen,
  }
}

export default AlphabetData
