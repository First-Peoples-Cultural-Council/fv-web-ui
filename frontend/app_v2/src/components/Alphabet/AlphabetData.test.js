import AlphabetData, { findSelectedCharacterData } from 'components/Alphabet/AlphabetData'

// Data
describe('AlphabetData', () => {
  const testCharacter = 'u'
  test('findSelectedCharacterData', () => {
    const output = findSelectedCharacterData({
      character: testCharacter,
      data: {
        characters: [
          {
            title: "k'",
            uid: 'e1134fc0-e4c4-4684-89d8-fb7ee1bd9bff',
            src: 'nxfile/default/b762405b-f98a-4b04-bbd2-1f710eaca9a0/file:content/En-us-river.ogg.mp3',
            relatedEntries: [
              {
                uid: '1-2-3',
                title: 'RelatedWord',
                definitions: ['defn1', 'defn2'],
                src:
                  'https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3',
              },
            ],
          },
          {
            title: 'v',
            uid: 'c6fa2b99-4076-43bb-adf9-9271ae7363d7',
            relatedEntries: [
              {
                uid: '1-2-3',
                title: 'RelatedWord',
                definitions: ['defn1', 'defn2'],
                src:
                  'https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3',
              },
            ],
          },
          {
            title: 'u',
            uid: '549cd27f-c0db-424d-8409-bf3070fc03f2',
            relatedEntries: [
              {
                uid: '1-2-3',
                title: 'RelatedWord',
                definitions: ['defn1', 'defn2'],
                src:
                  'https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3',
              },
            ],
          },
        ],
      },
      language: 'jestLanguage',
    })
    expect(output.title).toBe(testCharacter)
  })

  test('Ensure output is consistent', () => {
    const output = AlphabetData()
    expect(output).toMatchSnapshot()
  })
})
