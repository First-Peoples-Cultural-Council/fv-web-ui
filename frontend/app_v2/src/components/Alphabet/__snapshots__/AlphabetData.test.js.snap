// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlphabetData Ensure output is consistent 1`] = `
Object {
  "data": Array [
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
          "url": "/k'w/word/1-2-3",
        },
      ],
      "src": "/nuxeo/nxfile/default/b762405b-f98a-4b04-bbd2-1f710eaca9a0/file:content/En-us-river.ogg.mp3",
      "title": "k'",
      "uid": "e1134fc0-e4c4-4684-89d8-fb7ee1bd9bff",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "v",
      "uid": "c6fa2b99-4076-43bb-adf9-9271ae7363d7",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "u",
      "uid": "549cd27f-c0db-424d-8409-bf3070fc03f2",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "Ǭ",
      "uid": "054c4507-7cbc-4029-8a91-95924536d3a5",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "t",
      "uid": "7bc00011-9033-4aa7-8576-92c24915e6a5",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "Ǒ",
      "uid": "7774cc25-798a-43f5-9456-15e641f89b83",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ᐢ",
      "uid": "149bd537-8266-49cc-bf44-df43cac45abe",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "Ǔ",
      "uid": "a31726cc-e082-4e65-9ba9-a237c063a33c",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "y",
      "uid": "cc36139d-975b-4e16-aa30-b6ec3c446df9",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ᔦ",
      "uid": "12f30f26-4695-4118-bf5e-4826d3fa3164",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ᓱ",
      "uid": "7843ea62-c3c8-4a00-ac35-ecfa85f79ae5",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ḵ'",
      "uid": "37e070e8-4ffa-4bc7-b9c4-9a177071ada3",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "w",
      "uid": "8d90793c-b19b-46bd-9933-1d38748599db",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "s",
      "uid": "a14fa5a4-c628-42ae-8a29-f3d1911b5351",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "b",
      "uid": "9ec00e36-c836-4ef4-a98c-3e549c2925ce",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "Ǐ",
      "uid": "374dae5e-3806-4020-b7b9-a34a05b8df8c",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "x",
      "uid": "10f2af20-b065-4d31-8ccd-842dc2555d2e",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "a",
      "uid": "96adb6aa-56cc-45cf-b3ed-a525d7b40f93",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǳ",
      "uid": "92649908-f378-4644-acc7-59efd055fc94",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǂ",
      "uid": "16934d1b-49d7-4e55-a0d9-d1f82328c5fe",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǫ",
      "uid": "049720c3-c464-4be4-a27c-22e1f260b0d5",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǰ",
      "uid": "b58ac281-234e-4986-9d95-62e120da56bf",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ᔭ",
      "uid": "8e50f789-e454-423b-a472-7b20e182001c",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǧ",
      "uid": "14738ecf-c609-42da-a477-82096a691e0e",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ǎ",
      "uid": "78d5f6fb-c66e-403e-9aae-4d8ad1afd4fa",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "z",
      "uid": "cfcfded5-ea1f-4254-aa5f-c755cec18d2a",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "hl",
      "uid": "a608079e-b8f0-4942-931d-af1f1930009f",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "gw",
      "uid": "48ead292-8827-433b-a4a2-37bda6035d47",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "yy",
      "uid": "baae73af-00a6-46e8-817c-57584b96e06a",
    },
    Object {
      "relatedEntries": Array [
        Object {
          "definitions": Array [
            "defn1",
            "defn2",
          ],
          "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
          "title": "RelatedWord",
          "uid": "1-2-3",
        },
      ],
      "title": "ᓯ",
      "uid": "1dcac9c5-2fbd-4cf3-bae0-37e65ce00465",
    },
  ],
  "error": null,
  "isLoading": false,
  "language": "k'w",
  "links": Array [
    Object {
      "title": "Download Alphabet Pronunciation Guide",
      "url": "/url/1",
    },
    Object {
      "title": "Another potential related link",
      "url": "/url/2",
    },
  ],
  "onVideoClick": [Function],
  "selectedData": Object {
    "relatedEntries": Array [
      Object {
        "definitions": Array [
          "defn1",
          "defn2",
        ],
        "src": "https://dev.firstvoices.com/nuxeo/nxfile/default/15761975-8249-400c-a9e2-4a5357ca7ce9/file:content/sample1.mp3",
        "title": "RelatedWord",
        "uid": "1-2-3",
        "url": "/k'w/word/1-2-3",
      },
    ],
    "src": "/nuxeo/nxfile/default/b762405b-f98a-4b04-bbd2-1f710eaca9a0/file:content/En-us-river.ogg.mp3",
    "title": "k'",
    "uid": "e1134fc0-e4c4-4684-89d8-fb7ee1bd9bff",
  },
  "videoIsOpen": false,
}
`;
